### PR TITLE
Add support for PHP 8, fix tests

### LINKED ISSUE #55: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by ocean**
Targeting issue #55 

Still some test errors because of needing to change testing
for private/protected properties with PHPUnit 9 - I'm not sure how to test these properties now.

Signed-off-by: Drew Robinson <drew.robinson@gmail.com>

### CODE REVIEW

File: test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php [hunk 1]
```diff
@@ -14,17 +14,19 @@
 use Laminas\ServiceManager\AbstractFactory\ReflectionBasedAbstractFactory;
 use Laminas\ServiceManager\Exception\ServiceNotFoundException;
 use PHPUnit\Framework\TestCase;
-use Prophecy\Prophecy\ObjectProphecy;
```
Comment (froschdesign) [inline test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php:17]: The class is still in use. See the DocBlock in line 24.

Comment (ocean) [reply test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php:17]: Ok, added this back in.

**Comment by ocean** [comment]
The problematic area I need help with is around `/src/AbstractFactory/ReflectionBasedAbstractFactory.php:222`

**Comment by snapshotpl** [comment]
@ocean this patch helps for me:

```
--- a/src/AbstractFactory/ReflectionBasedAbstractFactory.php
+++ b/src/AbstractFactory/ReflectionBasedAbstractFactory.php
@@ -217,13 +217,13 @@ class ReflectionBasedAbstractFactory implements AbstractFactoryInterface
             return [];
         }
 
-        if ($parameter->hasType() && ! $parameter->getType()->isBuiltin()) {
+        if (! $parameter->hasType() || ($parameter->hasType() && $parameter->getType()->isBuiltin())) {
             if (! $parameter->isDefaultValueAvailable()) {
                 throw new ServiceNotFoundException(sprintf(
                     'Unable to create service "%s"; unable to resolve parameter "%s" '
                     . 'to a class, interface, or array type',
                     $requestedName,
-                    $parameter->getType()
+                    $parameter->getName()
                 ));
             }
```

**Comment by ocean** [comment]
@froschdesign This is ready for review now 😃

File: src/AbstractFactory/ReflectionBasedAbstractFactory.php [hunk 2]
```diff
@@ -249,4 +250,30 @@ private function resolveParameter(ReflectionParameter $parameter, ContainerInter
...
+        /**
+     * Gets the parameter's class name.
+     *
+     * @param ReflectionParameter $parameter
+     *   The parameter.
+     *
+     * @return string|null
+     *   The parameter's class name or NULL if the parameter is not a class.
+     */
+    public static function getParameterClassName(ReflectionParameter $parameter)
```
Comment (froschdesign) [inline src/AbstractFactory/ReflectionBasedAbstractFactory.php]: This method is declared as static, but is never used in this form. And I see no reason to extend the public API here.
Please set the visibility to private and remove `static`.

File: src/Tool/FactoryCreator.php [hunk 3]
```diff
@@ -80,6 +80,32 @@ public function createFactory($className)
...
+    /**
+     * Gets the parameter's class name.
+     *
+     * @param ReflectionParameter $parameter
+     *   The parameter.
+     *
+     * @return string|null
+     *   The parameter's class name or NULL if the parameter is not a class.
+     */
+    public static function getParameterClassName(ReflectionParameter $parameter)
```
Comment (froschdesign) [inline src/Tool/FactoryCreator.php]: The same here.

File: test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php [hunk 4]
```diff
@@ -171,7 +174,7 @@ public function testFactoryCanSupplyAMixOfParameterTypes()
         self::assertInstanceOf(TestAsset\ClassWithMixedConstructorParameters::class, $instance);
 
         self::assertEquals($config, $instance->config);
-        self::assertEquals([], $instance->options);
+        self::assertEquals(null, $instance->options);
```
Comment (froschdesign) [inline test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php]: Does this change mean that there was an error before?

**Review by froschdesign — CHANGES_REQUESTED** [review]
Thanks for your contribution! 👍 

Two changes are necessary and I have added a question.

File: test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php (hunk 4 above)
Comment (ocean) [reply test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php]: Yes, there was an error in the tests when I ran them locally. I thought I saw a comment something about `$options` always being treated as `null`, but now I can't find it. I'll revert this change and then you can see the error that shows up in Travis.

Comment (ocean) [reply test/AbstractFactory/ReflectionBasedAbstractFactoryTest.php]: @froschdesign Yep, in the test the `options` instance property is set to `null`, not to an empty array, see:
https://github.com/laminas/laminas-servicemanager/blob/3.5.x/test/AbstractFactory/TestAsset/ClassWithMixedConstructorParameters.php#L24

So either the assertion or the test class constructor needs changing I guess.

**Comment by boesing** [comment]
@ocean Is there anything we can help with? Feel free to ping me.

**Update: I've merged #56 which already had phpunit 9 fixes and thus your tests might pass after rebasing against 3.5.x branch 👍**

**Comment by boesing** [comment]
Oh, I wasnt aware that there was already #51 which added PHP 8.0 compatibility back in july.
I'll close this in favor of #51 as its tests are already passing.

**Comment by FabianKoestring** [comment]
Could you tag a new release for this feature?

**Comment by Ocramius** [comment]
@FabianKoestring done

**Comment by Ocramius** [comment]
@FabianKoestring beware: due to #59, this has been reverted for now, and will therefore be delayed.

**Pull request closed without merging**